### PR TITLE
[tune] Use `Checkpoint.to_bytes()` for store_to_object

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We currently use our own serialization to ship checkpoints as objects. Instead we should use the `Checkpoint` class. This PR also adds support to create results from checkpoints pointing to object references.

Depends on #26351

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — APPROVED** [review]
Thanks!

**Review by ArturNiederfahrenhorst — APPROVED** [review]

File: python/ray/tune/examples/xgboost_dynamic_resources_example.py [hunk 1]
```diff
@@ -22,12 +22,18 @@
...
 def get_best_model_checkpoint(analysis):
     best_bst = xgb.Booster()
-    try:
-        with open(analysis.best_checkpoint, "rb") as inputFile:
-            _, _, raw_model = pickle.load(inputFile)
-        best_bst.load_model(bytearray(raw_model))
-    except IsADirectoryError:
-        best_bst.load_model(os.path.join(analysis.best_checkpoint, CHECKPOINT_FILENAME))
+
+    with analysis.best_checkpoint.as_directory() as checkpoint_dir:
```
Comment (xwjiang2010) [inline python/ray/tune/examples/xgboost_dynamic_resources_example.py:26]: why does a user have to go through this to get best model checkpoint?
This seems too complicated to showcase as user code.

Comment (krfricke) [reply python/ray/tune/examples/xgboost_dynamic_resources_example.py:26]: This is for testing with the class trainable cc @Yard1 to maybe adjust this test in the future (e.g. store a separate `model.xgb` in the class trainable) - but likely out of scope for this PR which just adjusts existing logic to the new API.

File: python/ray/tune/execution/ray_trial_executor.py [hunk 2]
```diff
@@ -794,7 +795,8 @@ def restore(self, trial: Trial) -> None:
                 # This provides FT backwards compatibility in the
                 # case where no cloud checkpoints are provided.
                 logger.debug("Trial %s: Reading checkpoint into memory", trial)
-                obj = TrainableUtil.checkpoint_to_object(checkpoint_dir)
+                checkpoint_path = TrainableUtil.find_checkpoint_dir(checkpoint_dir)
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/execution/ray_trial_executor.py:798]: We should keep it for backwards compatibility with existing checkpoints. But yes, generally we can simplify this in the future

File: python/ray/tune/trainable/trainable.py [hunk 3]
```diff
@@ -568,25 +599,37 @@ def restore(self, checkpoint_path: str, checkpoint_node_ip: Optional[str] = None
                 f"Got checkpoint path: {checkpoint_path} and IP {checkpoint_node_ip}"
             )
 
-        with open(checkpoint_path + ".tune_metadata", "rb") as f:
-            metadata = pickle.load(f)
+        checkpoint_path = TrainableUtil.find_checkpoint_dir(checkpoint_path)
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:602]: should we be consistent with when we say `checkpoint_dir` vs `checkpoint_path`?
with the introduction of `relative_checkpoint_path`, seems we can achieve that?
is it actually checkpoint_dir here?

File: python/ray/tune/trainable/trainable.py [hunk 4]
```diff
@@ -428,26 +441,44 @@ def save(self, checkpoint_dir: Optional[str] = None) -> str:
...
+        # User saves checkpoint
         checkpoint_dict_or_path = self.save_checkpoint(checkpoint_dir)
-        trainable_state = self.get_state()
-        checkpoint_path = TrainableUtil.process_checkpoint(
-            checkpoint_dict_or_path,
-            parent_dir=checkpoint_dir,
-            trainable_state=trainable_state,
-        )
+
+        if checkpoint_dict_or_path is None:
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/trainable/trainable.py:454]: and path actually mean dir seems?

File: python/ray/tune/trainable/trainable.py [hunk 5]
```diff
@@ -428,26 +441,44 @@ def save(self, checkpoint_dir: Optional[str] = None) -> str:
             checkpoint_dir: Optional dir to place the checkpoint.
 
         Returns:
-            str: path that points to xxx.pkl file.
+            The given or created checkpoint directory.
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/trainable.py:444]: Nice!

File: python/ray/tune/trainable/function_trainable.py [hunk 6]
```diff
@@ -454,64 +453,57 @@ def step(self):
...
         checkpoint = self._status_reporter.get_checkpoint()
-        state = self.get_state()
 
         if not checkpoint:
-            state.update(iteration=0, timesteps_total=0, episodes_total=0)
             # We drop a marker here to indicate that the checkpoint is empty
             checkpoint = FuncCheckpointUtil.mk_null_checkpoint_dir(self.logdir)
             parent_dir = checkpoint
         elif isinstance(checkpoint, dict):
-            parent_dir = TrainableUtil.make_checkpoint_dir(
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/function_trainable.py:470]: if i am reading the code correctly, `make_checkpoint_dir ` is not returning anything now?

File: python/ray/tune/trainable/function_trainable.py [hunk 7]
```diff
@@ -454,64 +453,57 @@ def step(self):
...
 
+    def get_state(self):
+        state = super().get_state()
+
+        checkpoint = self._status_reporter.get_checkpoint()
+        if not checkpoint:
+            state.update(iteration=0, timesteps_total=0, episodes_total=0)
+        return state
+
     def save_checkpoint(self, tmp_checkpoint_dir: str = ""):
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/function_trainable.py:464]: change name to be consistent with trainable.py?

File: python/ray/tune/trainable/util.py [hunk 8]
```diff
@@ -19,60 +17,29 @@
...
-            trainable_state["saved_as_dict"] = saved_as_dict
-            pickle.dump(trainable_state, f)
-        return checkpoint_path
+    @staticmethod
+    def load_metadata(checkpoint_dir: str) -> Dict:
+        with open(os.path.join(checkpoint_dir, _TUNE_METADATA_FILENAME), "rb") as f:
+            return pickle.load(f)
 
     @staticmethod
     def load_checkpoint_metadata(checkpoint_path: str) -> Optional[Dict]:
```
Comment (xwjiang2010) [inline python/ray/tune/trainable/util.py:36]: Is this still needed?

File: python/ray/tune/trainable/trainable.py [hunk 9]
```diff
@@ -58,6 +60,8 @@
 
 SETUP_TIME_THRESHOLD = 10
 
+_METADATA_KEY = "_tune_metadata"
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:63]: No, leftover from previous iteration! Thanks!

File: python/ray/tune/trainable/util.py (hunk 8 above)
Comment (krfricke) [reply python/ray/tune/trainable/util.py:36]: Nope! Removed

File: python/ray/tune/trainable/trainable.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/trainable/trainable.py:602]: Yes I agree - from this point it's `checkpoint_dir`. Updated

**Review by xwjiang2010 — APPROVED** [review]

**Pull request merged**